packages/protocol/engine/tangle/booker/attachments.go

```diff
@@ -1,36 +1,111 @@
 }
 
 func newAttachments() (newAttachments *attachments) {
 	return &attachments{
-		attachments: memstorage.New[utxo.TransactionID, *memstorage.Storage[epoch.Index, *LockableSlice[*Block]]](),
-		evictionMap: memstorage.New[epoch.Index, set.Set[utxo.TransactionID]](),
+		attachments:        memstorage.New[utxo.TransactionID, *memstorage.Storage[epoch.Index, *memstorage.Storage[models.BlockID, *Block]]](),
+		evictionMap:        memstorage.New[epoch.Index, set.Set[utxo.TransactionID]](),
+		nonOrphanedCounter: memstorage.New[utxo.TransactionID, uint32](),
+
+		mutex: syncutils.NewDAGMutex[utxo.TransactionID](),
 	}
 }
 
-func (a *attachments) Store(txID utxo.TransactionID, block *Block) {
-	a.storeAttachment(txID, block)
+func (a *attachments) Store(txID utxo.TransactionID, block *Block) (created bool) {
+	a.mutex.Lock(txID)
+	defer a.mutex.Unlock(txID)
+
+	if !a.storeAttachment(txID, block) {
+		return false
+	}
+
+	prevValue, _ := a.nonOrphanedCounter.RetrieveOrCreate(txID, func() uint32 { return 0 })
+	a.nonOrphanedCounter.Set(txID, prevValue+1)
+
 	a.updateEvictionMap(block.ID().EpochIndex, txID)
+
+	return true
+}
+
+func (a *attachments) AttachmentOrphaned(txID utxo.TransactionID, block *Block) (attachmentBlock *Block, attachmentOrphaned, lastAttachmentOrphaned bool) {
+	a.mutex.Lock(txID)
+	defer a.mutex.Unlock(txID)
+
+	storage := a.storage(txID, false)
+	if storage == nil {
+		// zero attachments registered
+		return nil, false, false
+	}
+
+	attachmentsOfEpoch, exists := storage.Get(block.ID().Index())
+	if !exists {
+		// there is no attachments of the transaction in the epoch, so no need to continue
+		return nil, false, false
+	}
+
+	attachmentBlock, attachmentExists := attachmentsOfEpoch.Get(block.ID())
+	if !attachmentExists {
+		return nil, false, false
+	}
+
+	prevValue, counterExists := a.nonOrphanedCounter.Get(txID)
+	if !counterExists {
+		panic(fmt.Sprintf("non orphaned attachment counter does not exist for TxID(%s)", txID.String()))
+	}
+
+	newValue := prevValue - 1
+	a.nonOrphanedCounter.Set(txID, newValue)
+
+	return attachmentBlock, true, newValue == 0
 }
 
 func (a *attachments) Get(txID utxo.TransactionID) (attachments []*Block) {
 	if txStorage := a.storage(txID, false); txStorage != nil {
-		txStorage.ForEach(func(_ epoch.Index, blocks *LockableSlice[*Block]) bool {
-			attachments = append(attachments, blocks.Slice()...)
+		txStorage.ForEach(func(_ epoch.Index, blocks *memstorage.Storage[models.BlockID, *Block]) bool {
+			blocks.ForEach(func(_ models.BlockID, block *Block) bool {
+				attachments = append(attachments, block)
+				return true
+			})
+			return true
+		})
+	}
+
+	return
+}
+
+func (a *attachments) GetAttachmentBlocks(txID utxo.TransactionID) (attachments *set.AdvancedSet[*Block]) {
+	attachments = set.NewAdvancedSet[*Block]()
+
+	if txStorage := a.storage(txID, false); txStorage != nil {
+		txStorage.ForEach(func(_ epoch.Index, blocks *memstorage.Storage[models.BlockID, *Block]) bool {
```

Comment:
Why we don't lock the `mutex` DAGMutex for reading but we do only for writing?

Reply:
txStorage has a lock itself. We lock when writing because we had to store it and update the counters atomically.